lib/reversetunnel/agent.go

```diff
@@ -22,502 +22,533 @@ package reversetunnel
-		go a.handleGlobalRequests(a.ctx, reqs)
+	// Wait for open tranports to close before closing the connection.
+	a.drainWG.Wait()
 
-		return client, nil
+	a.cancel()
+	if a.client != nil {
+		a.client.Close()
 	}
-	return nil, trace.BadParameter("failed to dial: all auth methods failed")
+
+	a.wg.Wait()
+	return nil
 }
 
 // handleGlobalRequests processes global requests from the server.
-func (a *Agent) handleGlobalRequests(ctx context.Context, requestCh <-chan *ssh.Request) {
+func (a *agent) handleGlobalRequests(ctx context.Context, requests <-chan *ssh.Request) error {
 	for {
 		select {
-		case r := <-requestCh:
-			// When the channel is closing, nil is returned.
+		case r := <-requests:
+			// The request will be nil when the request channel is closing.
 			if r == nil {
-				return
+				return trace.Errorf("global request channel is closing")
 			}
 
 			switch r.Type {
 			case versionRequest:
-				err := r.Reply(true, []byte(teleport.Version))
+				err := a.client.Reply(r, true, []byte(teleport.Version))
 				if err != nil {
-					log.Debugf("Failed to reply to %v request: %v.", r.Type, err)
+					a.log.Debugf("Failed to reply to %v request: %v.", r.Type, err)
 					continue
 				}
 			default:
 				// This handles keep-alive messages and matches the behaviour of OpenSSH.
-				err := r.Reply(false, nil)
+				err := a.client.Reply(r, false, nil)
 				if err != nil {
-					log.Debugf("Failed to reply to %v request: %v.", r.Type, err)
+					a.log.Debugf("Failed to reply to %v request: %v.", r.Type, err)
 					continue
 				}
 			}
 		case <-ctx.Done():
-			return
+			return trace.Wrap(ctx.Err())
 		}
 	}
 }
 
-// run is the main agent loop. It tries to establish a connection to the
-// remote proxy and then process requests that come over the tunnel.
-//
-// Once run connects to a proxy it starts processing requests from the proxy
-// via SSH channels opened by the remote Proxy.
-//
-// Agent sends periodic heartbeats back to the Proxy and that is how Proxy
-// determines disconnects.
-func (a *Agent) run() {
-	defer a.setState(agentStateDisconnected)
-	defer a.Lease.Release()
-
-	a.setState(agentStateConnecting)
-
-	// Try and connect to remote cluster.
-	conn, err := a.connect()
-	if err != nil || conn == nil {
-		a.log.Warningf("Failed to create remote tunnel: %v, conn: %v.", err, conn)
-		return
+func (a *agent) isDraining() bool {
+	select {
+	case <-a.drainCtx.Done():
+		return true
+	default:
+		return false
 	}
+}
 
-	local := conn.LocalAddr().String()
-	remote := conn.RemoteAddr().String()
-	defer func() {
-		if err := conn.Close(); err != nil {
-			a.log.Warnf("Failed to close remote tunnel: %v, local addr: %s remote addr: %s", err, local, remote)
-		}
+// signalDraining will signal one time when the draining context is canceled.
+func (a *agent) signalDraining() <-chan struct{} {
+	c := make(chan struct{})
+	a.wg.Add(1)
+	go func() {
+		<-a.drainCtx.Done()
+		c <- struct{}{}
```

Comment:
you could alternatively just close the channel

Reply:
Fixed this and other feedback here [31adbca](https://github.com/gravitational/teleport/pull/10444/commits/31adbca9fd58d548b0a488e0681c6dde416721bb)